@xen-orchestra/fs/src/azure.js

```diff
@@ -0,0 +1,195 @@
+        this[functionName] = pRetry.wrap(this[functionName], {
+          delays: [100, 200, 500, 1000, 2000],
+          when: err => !['EEXIST', 'EISDIR', 'ENOTEMPTY', 'ENOENT', 'ENOTDIR', 'EISDIR'].includes(err?.code),
+          onRetry(error) {
+            warn('retrying method on fs ', {
+              method: functionName,
+              attemptNumber: this.attemptNumber,
+              delay: this.delay,
+              error,
+              file: this.arguments?.[0],
+            })
+          },
+        })
+      }
+    })
+  }
+
+  get type() {
+    return 'azure'
+  }
+
+  async #createContainer() {
+    await this.#containerClient.createIfNotExists()
+  }
+
+  async #makePrefix(path) {
+    return path.endsWith('/') ? path : `${path}/`
+  }
+
+  async #isNotEmptyDir(dir) {
+    const prefix = this.#makePrefix(dir)
+    const blobs = []
+    for await (const blob of this.#containerClient.listBlobsFlat({ prefix })) {
+      blobs.push(blob.name)
+    }
+
+    return blobs.length > 0
+  }
+
+  async #streamToBuffer(readableStream, buffer) {
+    return new Promise((resolve, reject) => {
+      let bytesRead = 0
+
+      readableStream.on('data', data => {
+        if (bytesRead + data.length <= buffer.length) {
+          data.copy(buffer, bytesRead)
+          bytesRead += data.length
+        } else {
+          reject(new Error('Buffer size exceeded'))
+        }
+      })
+      readableStream.on('end', () => {
+        resolve(bytesRead)
+      })
+      readableStream.on('error', reject)
+    })
+  }
+
+  // this can be used right away instead of _writeFile func
+  async #_outputStream(file, data) {
+    const blobClient = this.#containerClient.getBlockBlobClient(file)
+    await blobClient.uploadStream(data, MAX_BLOCK_SIZE, MAX_BLOCK_COUNT)
+  }
+
+  // list containers
+  async _listContainers() {
+    const containers = []
+    for await (const container of this.#blobServiceClient.listContainers()) {
+      containers.push(container.name)
+    }
+    return containers
+  }
+
+  // list blobs in container
+  async _list() {
+    const files = []
+    for await (const blob of this.#containerClient.listBlobsFlat(this.#container)) {
+      files.push(basename(blob.name))
+    }
+    return files
+  }
+
+  // uploads a file to a blob
+  async _writeFile(file, data) {
+    const blobClient = this.#containerClient.getBlockBlobClient(file)
+    if (data.length > MAX_BLOCK_SIZE) {
+      await this.#_outputStream(file, data)
+    } else {
+      await blobClient.upload(data, data.length)
+    }
+  }
+
+  async _createReadStream() {
+    try {
+      const blobClient = this.#containerClient.getBlobClient(this.#dir)
+      const response = await blobClient.download()
+      return await response.readableStreamBody
+    } catch (e) {
+      if (e.name === 'NoSuchKey') {
+        const error = new Error(`ENOENT: no such file '${this.#dir}'`)
+        error.code = 'ENOENT'
+        error.path = this.#dir
+        throw error
+      }
```

Comment:
this is the S3 specific error handling , you must adapt this code to azure ( blob not found) and emit a ENOENT error with the right parameter